File: Lab 5/binary.c
#include <stdio.h>

int main() {
    int n, target, left, right, mid, step = 1;
    
    //Taking array size input
    printf("Enter the number of elements: ");
    scanf("%d", &n);

    int arr[n];

    //Taking sorted array input
    printf("Enter %d sorted elements: ", n);
    for (int i = 0; i < n; i++) {
        scanf("%d", &arr[i]);
    }

    //Taking target element input
    printf("Enter the number to search: ");
    scanf("%d", &target);

    //Binary Search Logic
    left = 0;right = n - 1;

    printf("\nBinary Search Process:\n");

    while (left <= right) {
        mid = left + (right - left) / 2;

        //Print the current step details
        printf("Step %d: Searching in range [%d - %d], Mid = arr[%d] = %d\n", step, left+1, right+1, mid+1, arr[mid]);
        printf("Step %d: Comparing target %d with arr[%d] = %d\n", step, target, mid+1, arr[mid]);

        if (arr[mid] == target) {
            printf("Target %d found at possition %d\n", target, mid+1);
            return 0;
        } else if (arr[mid] < target) {
            printf("Target is greater than arr[%d], searching right half\n", mid);
            left = mid + 1;
        } else {
            printf("Target is less than arr[%d], searching left half\n", mid);
            right = mid - 1;
        }

        step++;
    }

    printf("Target %d not found in the array\n", target);

    return 0;
}

File: Lab 5/binary.py
def main():
    #Taking array size input
    n = int(input("Enter the number of elements: "))

    #Taking sorted array input
    arr = list(map(int, input(f"Enter {n} sorted elements: ").split()))

    #Taking target element input
    target = int(input("Enter the number to search: "))

    #Binary Search
    left = 0
    right = n - 1
    step = 1

    print("\nBinary Search Process:")

    while left <= right:
        mid = left + (right - left) // 2

        #Print the current step details
        print(f"Step {step}: Searching in range [{left+1} - {right+1}], Mid = arr[{mid+1}] = {arr[mid]}")
        print(f"Step {step}: Comparing target {target} with arr[{mid+1}] = {arr[mid]}")

        if arr[mid] == target:
            print(f"Target {target} found at position {mid + 1}")
            return
        elif arr[mid] < target:
            print(f"Target is greater than arr[{mid+1}], searching right half")
            left = mid + 1
        else:
            print(f"Target is less than arr[{mid+1}], searching left half")
            right = mid - 1

        step = step + 1

    print(f"Target {target} not found in the array")
  
if __name__ == "__main__":
    main()
